File: tests/fixtures/call_graph_sample.py
"""
Test sample with complex call graph for Phase 2 testing
Contains:
- Multi-level call relationships
- Recursive calls
- Cross-file call references
"""


def helper_function(x):
    """A helper function called by multiple functions"""
    return x * 2


def recursive_factorial(n):
    """Recursive factorial function"""
    if n <= 1:
        return 1
    return n * recursive_factorial(n - 1)


def fibonacci(n):
    """Fibonacci function using helper"""
    if n <= 1:
        return n
    return helper_function(fibonacci(n - 1) + fibonacci(n - 2))


def process_data(data):
    """Main processing function with multi-level calls"""
    processed = []

    for item in data:
        # Call multiple functions
        item = helper_function(item)
        if item > 10:
            item = recursive_factorial(min(item, 5))
        processed.append(item)

    return processed


def main_entry():
    """Entry point that orchestrates all functions"""
    data = [1, 2, 3, 4, 5]
    result = process_data(data)
    result.append(fibonacci(10))
    return result

File: tests/fixtures/call_graph_sample.ts
/**
 * Test sample with complex call graph for Phase 2 testing
 * Contains:
 * - Multi-level call relationships
 * - Method calls on objects
 * - Chained calls
 */


interface User {
  id: number;
  name: string;
  email: string;
}


export class UserService {
  private users: User[] = [];

  addUser(user: User): void {
    this.users.push(user);
  }

  findUserById(id: number): User | null {
    // @ts-ignore
    return this.users.find(u => u.id === id) || null;
  }

  validateUser(user: User): boolean {
    return this.validateEmail(user.email) && this.validateName(user.name);
  }

  private validateEmail(email: string): boolean {
    // @ts-ignore
    return email.includes('@');
  }

  private validateName(name: string): boolean {
    return name.length > 0;
  }
}


export class AuthService {
  private userService: UserService;

  constructor(userService: UserService) {
    this.userService = userService;
  }

  authenticate(userId: number, token: string): boolean {
    const user = this.userService.findUserById(userId);
    if (!user) {
      return false;
    }
    return this.validateToken(token);
  }

  private validateToken(token: string): boolean {
    return token.length > 10;
  }

  login(userId: number): string {
    const token = this.generateToken();
    if (this.authenticate(userId, token)) {
      return token;
    }
    throw new Error('Authentication failed');
  }

  private generateToken(): string {
    return 'token_' + Math.random().toString(36).substring(2);
  }
}
